arbnode/maintenance.go

```diff
@@ -120,157 +72,89 @@ func NewMaintenanceRunner(config MaintenanceConfigFetcher, seqCoordinator *SeqCo
-	}
-}
-
-func (mr *MaintenanceRunner) setMaintenanceStart() error {
-	prev := mr.lastMaintenance.Swap(0)
-	if prev == 0 {
-		return errors.New("already running")
-	}
-	return nil
-}
-
-func (mr *MaintenanceRunner) maybeRunScheduledMaintenance(ctx context.Context) time.Duration {
+// exported for testing
+func (mr *MaintenanceRunner) MaybeRunMaintenance(ctx context.Context) time.Duration {
 	config := mr.config()
-	if !config.enabled {
-		return time.Minute
+	if !config.Enable {
+		log.Debug("maintenance is disabled, skipping")
+		return config.RunInterval
 	}
 
-	now := time.Now().UTC()
-
-	inMaintenance, lastMaintenance := mr.getPrevMaintenance()
-	if inMaintenance {
-		return time.Minute
-	}
-
-	if !wentPastTimeOfDay(lastMaintenance, now, config.minutesAfterMidnight) {
-		return time.Minute
-	}
-
-	err := mr.attemptMaintenance(ctx)
+	shouldTriggerMaintenance, err := mr.exec.ShouldTriggerMaintenance().Await(mr.GetContext())
 	if err != nil {
-		log.Warn("scheduled maintenance error", "err", err)
+		log.Error("error checking if maintenance should be triggered", "err", err)
+		return config.RunInterval
 	}
-
-	return time.Minute
-}
-
-func (mr *MaintenanceRunner) Trigger() error {
-	if !mr.config().Triggerable {
-		return errors.New("maintenance not configured to be triggerable")
-	}
-	if running, _ := mr.getPrevMaintenance(); running {
-		return nil
+	if !shouldTriggerMaintenance {
+		log.Debug("skipping maintenance, not triggered")
+		return config.RunInterval
 	}
-	// maintenance takes a long time, run on a separate thread
-	mr.LaunchThread(func(ctx context.Context) {
-		err := mr.attemptMaintenance(ctx)
-		if err != nil {
-			log.Warn("triggered maintenance returned error", "err", err)
-		}
-	})
-	return nil
-}
 
-func (mr *MaintenanceRunner) attemptMaintenance(ctx context.Context) error {
+	// If seqCoordinator is nil there is no need to coordinate maintenance running with other sequecers.
 	if mr.seqCoordinator == nil {
-		return mr.runMaintenance()
+		mr.runMaintenance()
+		return config.RunInterval
 	}
 
 	release := make(chan struct{})
 	if !mr.lock.AttemptLockAndPeriodicallyRefreshIt(ctx, release) {
-		return errors.New("did not catch maintenance lock")
+		log.Warn("maintenance lock not acquired, skipping maintenance")
+		return config.RunInterval
 	}
 	defer func() {
 		release <- struct{}{}
 	}()
 
-	res := errors.New("maintenance failed to hand-off chosen one")
-
-	log.Info("Attempting avoiding lockout and handing off", "targetTime", mr.config().TimeOfDay)
-	// Avoid lockout for the sequencer and try to handoff.
+	log.Info("Attempting avoiding lockout and handing off")
 	if mr.seqCoordinator.AvoidLockout(ctx) && mr.seqCoordinator.TryToHandoffChosenOne(ctx) {
-		res = mr.runMaintenance()
+		log.Info("Avoided lockout and handed off chosen one")
+		mr.runMaintenance()
+	} else {
+		log.Error("maintenance failed to hand-off chosen one")
+	}
+	defer mr.seqCoordinator.SeekLockout(ctx)
```

Comment:
defer here just before returning is weird

Reply:
Removed defer